engine/scene/componentdeserializer.h

```diff
@@ -0,0 +1,52 @@
+#pragma once
+
+#include <nlohmann/json.hpp>
+
+namespace pong
+{
+
+class Ball;
+class BaseComponent;
+class Behavior;
+class Button;
+class CheckBox;
+class ColliderBox;
+class Mesh;
+class Player;
+class Opponent;
+class ScoreArea;
+class ScoreController;
+class SettingsScreenController;
+class Slider;
+class Text;
+class TitleScreenController;
+class Transform;
```

Comment:
Seems like the visitor pattern is a little prone to inviting circular dependencies. Is there a better way to fix these? It is not the worst to forward declaring everything, but it is less clean.

Reply:
Yep, this version is known as the "cyclic visitor" and is specifically known for needing a circular reference. There is an "acylic visitor" but the performance cost comes into play with another direction.

There is a way around it using value semantics std::variant and std::visit, but it is a bit of a change.

He talks about it in detail here if you are curious:
https://youtu.be/PEcy1vYHb8A?si=y2rM3EAlMZUu8Aju



I think the forward declaration here is good. Not that it prevents things from being recompiled if something changes in the Opponent.h header.